spec/haskell/src/SEL4/Object/SchedContext.lhs

```diff
@@ -345,15 +345,46 @@ This module uses the C preprocessor to select a target architecture.
 > refillHeadOverlappingLoop scPtr =
 >        whileLoop (const (fromJust . runReaderT (refillHeadOverlapping scPtr))) (const (mergeRefills scPtr)) ()
 
+> headTimeBuffer :: Ticks -> KernelR Bool
+> headTimeBuffer usage = do
+>     scPtr <- readCurSc
+>     sc <- readSchedContext scPtr
+>     return $ rAmount (refillHd sc) <= usage && (maxBound :: Word64) - rTime (refillHd sc) >= 5 * usToTicks maxPeriodUs
+
+> handleOverrunLoopBody :: Ticks -> Kernel Ticks
+> handleOverrunLoopBody usage = do
+>     scPtr <- getCurSc
+>     single <- refillSingle scPtr
+>     sc <- getSchedContext scPtr
+
+>     usage' <- return (usage - rAmount (refillHd sc))
+
+>     if single
+>        then updateRefillHd scPtr $ \r -> r { rTime = rTime r + scPeriod sc}
+>        else do
+>          old_head <- refillPopHead scPtr
+>          let new = old_head { rTime = rTime old_head + scPeriod sc}
+>          scheduleUsed scPtr new
+
+>     return usage'
+
+> handleOverrunLoop :: Ticks -> Kernel Ticks
+> handleOverrunLoop usage =
+>     whileLoop (\usage -> fromJust . runReaderT (headTimeBuffer usage)) (\usage -> handleOverrunLoopBody usage) usage
+
 > refillBudgetCheck :: Ticks -> Kernel ()
 > refillBudgetCheck usage = do
 >     scPtr <- getCurSc
```

Comment:
If we know that it's not round robin in the abstract or exec level, it's very straightforward to carry that down. Do we know that? I'm not sure what the setup for CRefine looks like. It's very easy to add to the Haskell, so I can certainly do that

Reply:
Well, having looked a bit further, you assert that it's not round robin in the abstract spec so you must have already proven it as part of refine. If that's true then we would want to add the assert to Haskell as well, so we can reuse it during crefine. Otherwise we would have to reprove it.

Reply:
The `corres` rules for `refillBudgeCheck` and `refillUnblockCheck` aren't done yet. That'll be fun. But OK, I'll add the assert